lib/protobuf/extype/extype_protocol.ex

```diff
@@ -0,0 +1,85 @@
+defprotocol Extype.Protocol do
+  # Imagine registering an extension by defining a protocol
+
+  @type extype :: atom
+
+  @typedoc """
+  The existing type of the field. Often the module name of the struct.
+  """
+  @type type :: atom
+
+  @typedoc """
+  A value with type extype.
+  """
+  @type value :: struct | any
+
+  @spec validate_and_to_atom_extype!(type, option :: String.t) :: atom
+  def validate_and_to_atom_extype!(type, option)
+
+  @spec do_type_default(type, extype) :: any
+  def do_type_default(type, extype)
+
+  @spec do_new(type, value, extype) :: value
+  def do_new(type, value, extype)
+
+  @spec do_encode_type(type, value, extype) :: binary
+  def do_encode_type(type, v, extype)
+
+  @spec do_decode_type(type, val :: binary, extype) :: value
+  def do_decode_type(val, type, extype)
+end
+
+defmodule Extype do
+  @moduledoc "Extype"
+
+  @type extype :: Extype.Protocol.extype
+  @type type :: Extype.Protocol.type
+  @type value :: Extype.Protocol.value
+
+  # A serious trick
+  def get_mod(type) when is_atom(type) do
+    try do
+      Extype.Protocol.impl_for!(%{__struct__: type})
```

Comment:
You could potentially have a default implementation of `Extype.Protocol` for `Any` and add a `supported?` function that returns false for that implementation. Not sure what's better.

Reply:
I'm going to leave it this way for now but we can think about making it bettter v2